File: src/main/scala/com/yogpc/qp/integration/jei/MoverRecipeCategory.scala
package com.yogpc.qp.integration.jei

import java.util.Collections

import com.mojang.blaze3d.matrix.MatrixStack
import com.yogpc.qp.machines.base.IEnchantableItem
import com.yogpc.qp.utils.Holder
import com.yogpc.qp.{QuarryPlus, _}
import mezz.jei.api.constants.VanillaTypes
import mezz.jei.api.gui.IRecipeLayout
import mezz.jei.api.gui.drawable.IDrawable
import mezz.jei.api.helpers.IGuiHelper
import mezz.jei.api.ingredients.IIngredients
import mezz.jei.api.recipe.category.IRecipeCategory
import net.minecraft.client.Minecraft
import net.minecraft.client.resources.I18n
import net.minecraft.enchantment.Enchantment
import net.minecraft.item.{Item, ItemStack, Items}
import net.minecraft.util.ResourceLocation
import net.minecraft.util.text.TranslationTextComponent
import net.minecraftforge.registries.ForgeRegistries

import scala.jdk.CollectionConverters._

class MoverRecipeCategory(guiHelper: IGuiHelper) extends IRecipeCategory[MoverRecipeCategory.MoverRecipe] {

  import MoverRecipeCategory._

  override def getUid: ResourceLocation = UID

  override def getTitle: String = I18n.format(Holder.blockMover.getTranslationKey)

  override def getBackground: IDrawable = guiHelper.createDrawable(backGround, xOff, yOff, 167, 76)

  override def setRecipe(recipeLayout: IRecipeLayout, recipeWrapper: MoverRecipe, ingredients: IIngredients): Unit = {
    val guiItemStack = recipeLayout.getItemStacks

    guiItemStack.init(0, true, 3, 30)
    guiItemStack.init(1, false, 3 + 144, 30)

    guiItemStack.set(ingredients)
  }

  override def getRecipeClass: Class[_ <: MoverRecipe] = classOf[MoverRecipe]

  override def getIcon: IDrawable = guiHelper.createDrawableIngredient(new ItemStack(Holder.blockMover))

  override def setIngredients(recipe: MoverRecipe, ingredients: IIngredients): Unit = {
    val input = recipe.enchantments.flatMap(e => pickaxes.map(i => e -> i.enchantmentAdded(e, e.getMaxLevel)))
    val output = input.map { case (e, _) => recipe.toStack.enchantmentAdded(e, e.getMaxLevel) }

    ingredients.setInputLists(VanillaTypes.ITEM, Collections.singletonList(input.map(_._2).asJava))
    ingredients.setOutputLists(VanillaTypes.ITEM, Collections.singletonList(output.asJava))
  }

  override def draw(recipe: MoverRecipe, matrixStack: MatrixStack, mouseX: Double, mouseY: Double): Unit = {
    super.draw(recipe, matrixStack, mouseX, mouseY)
    recipe.enchantments.zipWithIndex.foreach { case (e, i) =>
      Minecraft.getInstance().fontRenderer.func_238422_b_(matrixStack, new TranslationTextComponent(e.getName).func_241878_f(), (36 - xOff).toFloat, (6 - yOff + 10 * i).toFloat, 0x404040)
    }
  }
}

object MoverRecipeCategory {
  final val UID = new ResourceLocation(QuarryPlus.modID, "quarryplus.enchantmover")
  val backGround = new ResourceLocation(QuarryPlus.modID, "textures/gui/mover_jei.png")
  final val xOff = 0
  final val yOff = 0
  final val o = 18

  def recipes: java.util.Collection[MoverRecipe] =
    ForgeRegistries.ITEMS.asScala
      .collect { case i: Item with IEnchantableItem => i }
      .flatMap(item => item.stacks().map(s => MoverRecipe(item, s)))
      .toSeq.asJava

  case class MoverRecipe(item: Item with IEnchantableItem, stack: ItemStack) extends (Enchantment => Boolean) {

    import net.minecraft.enchantment.Enchantments._

    val enchantments: Seq[Enchantment] = Seq(EFFICIENCY, UNBREAKING, FORTUNE, SILK_TOUCH).filter(this)

    def toStack: ItemStack = stack.copy()

    def canMove(enchantment: Enchantment): Boolean = item.canMove(toStack, enchantment)

    override def apply(v1: Enchantment): Boolean = canMove(v1)

    override def toString(): String = s"MoverRecipe{$item, $enchantments}"
  }

  def pickaxes = List(new ItemStack(Items.DIAMOND_PICKAXE), new ItemStack(Items.NETHERITE_PICKAXE))
}

File: src/main/scala/com/yogpc/qp/integration/jei/MoverRecipeCategory.java
package com.yogpc.qp.integration.jei;

import java.util.List;

import com.mojang.blaze3d.vertex.PoseStack;
import com.yogpc.qp.Holder;
import com.yogpc.qp.QuarryPlus;
import com.yogpc.qp.machines.EnchantableItem;
import com.yogpc.qp.machines.EnchantmentLevel;
import mezz.jei.api.constants.VanillaTypes;
import mezz.jei.api.gui.builder.IRecipeLayoutBuilder;
import mezz.jei.api.gui.drawable.IDrawable;
import mezz.jei.api.gui.ingredient.IRecipeSlotsView;
import mezz.jei.api.helpers.IGuiHelper;
import mezz.jei.api.recipe.IFocusGroup;
import mezz.jei.api.recipe.RecipeIngredientRole;
import mezz.jei.api.recipe.RecipeType;
import mezz.jei.api.recipe.category.IRecipeCategory;
import net.minecraft.client.Minecraft;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.enchantment.Enchantment;
import net.minecraftforge.registries.ForgeRegistries;
import org.apache.commons.lang3.tuple.Pair;

class MoverRecipeCategory implements IRecipeCategory<MoverRecipeCategory.MoverRecipe> {
    public static final RecipeType<MoverRecipe> RECIPE_TYPE = RecipeType.create(QuarryPlus.modID, "quarryplus.enchantmover", MoverRecipe.class);
    private static final ResourceLocation backGround = new ResourceLocation(QuarryPlus.modID, "textures/gui/mover_jei.png");
    private static final int xOff = 0;
    private static final int yOff = 0;
    private final IGuiHelper helper;
    private final List<ItemStack> pickaxes;

    public MoverRecipeCategory(IGuiHelper helper) {
        this.helper = helper;
        this.pickaxes = List.of(new ItemStack(Items.DIAMOND_PICKAXE), new ItemStack(Items.NETHERITE_PICKAXE));
    }

    @Override
    public RecipeType<MoverRecipe> getRecipeType() {
        return RECIPE_TYPE;
    }

    @Override
    public Component getTitle() {
        return Holder.BLOCK_MOVER.getName();
    }

    @Override
    public IDrawable getBackground() {
        return helper.createDrawable(backGround, xOff, yOff, 167, 76);
    }

    @Override
    public IDrawable getIcon() {
        return helper.createDrawableIngredient(VanillaTypes.ITEM_STACK, new ItemStack(Holder.BLOCK_MOVER));
    }

    @Override
    public void setRecipe(IRecipeLayoutBuilder builder, MoverRecipe recipe, IFocusGroup focuses) {
        var input = recipe.makeInput(pickaxes);
        var output = input.stream().map(Pair::getKey).map(recipe::makeOutput).toList();

        builder.addSlot(RecipeIngredientRole.INPUT, 4, 31)
            .addItemStacks(input.stream().map(Pair::getValue).toList());
        builder.addSlot(RecipeIngredientRole.OUTPUT, 4 + 144, 31)
            .addItemStacks(output);
    }

    @Override
    public void draw(MoverRecipe recipe, IRecipeSlotsView recipeSlotsView, PoseStack stack, double mouseX, double mouseY) {
        var enchantments = recipe.item.acceptEnchantments().stream().map(e -> new EnchantmentLevel(e, 1))
            .sorted(EnchantmentLevel.QUARRY_ENCHANTMENT_COMPARATOR).map(EnchantmentLevel::enchantment).toList();
        for (int i = 0; i < enchantments.size(); i++) {
            var text = Component.translatable(enchantments.get(i).getDescriptionId());
            Minecraft.getInstance().font.draw(stack, text, 36 - xOff, 6 - yOff + 10 * i, 0x404040);
        }
    }

    record MoverRecipe(EnchantableItem item, ItemStack stack) {
        List<Pair<Enchantment, ItemStack>> makeInput(List<ItemStack> pickaxes) {
            return item.acceptEnchantments().stream()
                .flatMap(e -> pickaxes.stream().map(ItemStack::copy).peek(i -> i.enchant(e, e.getMaxLevel()))
                    .map(i -> Pair.of(e, i)))
                .toList();
        }

        ItemStack makeOutput(Enchantment enchantment) {
            var s = stack.copy();
            s.enchant(enchantment, enchantment.getMaxLevel());
            return s;
        }
    }

    static List<MoverRecipe> recipes() {
        return ForgeRegistries.ITEMS.getValues().stream()
            .filter(i -> i instanceof EnchantableItem)
            .map(i -> new MoverRecipe((EnchantableItem) i, new ItemStack(i)))
            .toList();
    }
}
